src/bidiMapper/domains/context/cdpTarget.ts

```diff
@@ -228,6 +230,47 @@ export class CdpTarget {
     }
   }
 
+  // TODO: Document.
+  async addPreloadScriptWithChannels(
+    functionDeclaration: string,
+    channels?: Script.ChannelValue[],
+    sandbox?: string
+  ): Promise<Script.PreloadScript> {
+    // TODO: Document.
+    const createScriptWithChannels = (
+      functionDeclaration: string,
+      channels?: Script.ChannelValue[]
+    ): string => {
+      if (channels === undefined) {
+        // The spec provides a function, and CDP expects an evaluation.
+        return `(${functionDeclaration})();`;
+      }
+
+      const uuid = uuidv4();
+
+      // See: http://go/chrome-devtools:bootstrap-bindings-proposal
+      const intermediateScript = `
+      const channelProxy = (${newChannel()})();
+      window['${uuid}'] = channelProxy.getMessage;
```

Comment:
Please add TODO to make it [non-enumerable](https://docs.google.com/document/d/1rI078gHjmR1cL1cjKtrGoDAsJoDVLkX_aF64pc8hgOs/edit#bookmark=id.3j1581qxll6v).

Reply:
In order to resist race condition between running the preload script and initialisation of long poll, you can put a promise in `window['${uuid}']` if its not set, or resolve the promise, if it is.

Reply:
```suggestion
      // We already tried to initialize long poll for this chanel:
      if(window['${uuid}']) window['${uuid}'].resolve(channelProxy.getMessage);
      else
           window['${uuid}'] = Proxy.resolve(channelProxy.getMessage);
```

Reply:
Drive-by question: instead of

```
window['${uuid}']
```

did you mean

```
window[`${uuid}`]
```

or more simply,

```
window[uuid]
```

?

Reply:
Ah, I missed that we’re in a template literal context already starting on line 252. We could use `JSON.stringify()` instead of hardcoding the `'` (although it shouldn’t matter for properly formed UUIDs).

Reply:
> 

This is logic of constructing a string which will be evaluated in the new document.